gh-98831: Support conditional effects; use for LOAD_ATTR

gvanrossum opened the pull request:
We can now write
```
inst(OP, (foo if (oparg & 1) -- bar if (oparg & 2)) {
    ...
}
```
which pops `foo` off the stack if `oparg & 1`, otherwise setting it to `NULL`, and pushes `bar` onto the stack if `oparg & 2` (otherwise ignoring it).

This syntax cannot be combined with an array size on the same effect, but it can be combined with a type (untested). In addition, it is incompatible with an array size closer to the stack top, e.g. `foo if (oparg&1), bar[oparg>>1]` is not supported (I don't think it's used, and generating code for it would be slightly awkward, but could be done if needed).

The implementation switches to using `POP()` and `PUSH()` (from `PEEK()` and `POKE()`) for those variables that are either conditional or whose position relative to the bottom (!) of the stack depends on a condition. (See the added test for details.)

To demonstrate this, I converted `LOAD_ATTR` (but not its specializations).

<!-- gh-issue-number: gh-98831 -->
* Issue: gh-98831
<!-- /gh-issue-number -->

gvanrossum committed:
Add parser support for conditional stack effects

(Semantics aren't there yet, and the test fails.)

gvanrossum committed:
Fix a typo in test_generator.py that made it into main :-(

gvanrossum committed:
Make conditional effects work, in theory

gvanrossum committed:
Convert LOAD_ATTR to new format

First use of conditional stack effects.

gvanrossum committed:
Conditionalize effect_size()

gvanrossum commented on Python/bytecodes.c:
```diff
@@ -1438,13 +1438,11 @@ dummy_func(
 
-        // error: LOAD_ATTR has irregular stack effect
-        inst(LOAD_ATTR) {
+        inst(LOAD_ATTR, (unused/9, owner -- res, res2 if (oparg & 1))) {
```
Check out my comment in the generated code about 'res2'.

gvanrossum commented on Python/generated_cases.c.h:
```diff
@@ -1745,11 +1745,13 @@
             PREDICTED(LOAD_ATTR);
+            PyObject *owner = PEEK(1);
+            PyObject *res;
+            PyObject *res2;
```
Should this initialize `res2 = NULL`? Right now I get compiler warnings (in GitHub) about a possibly uninitialized variable:
```
‘res2’ may be used uninitialized in this function [-Wmaybe-uninitialized]
```
presumably from the `PUSH()` on L1798.

iritkatriel replied:
I think the issue is that the compiler can't see that ``if (oparg & 1)`` is the same as ``if (_cond_res2)``, so it can't tell that res2 is always initialised when it is used.  Not sure what to do about it, add a syntax in the DSL to access the condition from within the bytecode definition?

gvanrossum replied:
I'll just initialize it to NULL. Perhaps some later phase of optimization will recognize that for the dead value it is.

iritkatriel replied:
I think if we rename `_cond_res2` to `res2_exists` and then in the body of the bytecode we do `if(res2_exists) /* use res2 */` instead of `if(oparg &1) /* use res2 */` then it would make the code clearer.  It's also safer, because if you change the args to the instr then `oparg&1` might not be correct anymore, but `res2_exists` will be automatically updated.

gvanrossum replied:
Okay, let's try that. I had a commit in the queue that initializes the variable to NULL, which I think would also get rid of the problem (if not I'd be very surprised) but since I didn't push that let's first try it with your idea.

gvanrossum replied:
Hm, looks like it doesn't help, the error is still there. So instead let me try this with the (much simpler) "= NULL" solution, which *has* to work. In fact, I may also drop the flag variable idea, and solve the warning about bool by writing `X ? 1 : 0` instead of `X != 0`.

gvanrossum commented on Python/generated_cases.c.h:
```diff
@@ -1788,20 +1782,21 @@
             }
-            JUMPBY(INLINE_CACHE_ENTRIES_LOAD_ATTR);
+            POKE(1, res);
+            if (oparg & 1) { PUSH(res2); }
```
Here (I think).

iritkatriel commented on Tools/cases_generator/generate_cases.py:
```diff
@@ -220,10 +228,23 @@ def __init__(self, inst: parser.InstDef):
+        manual = False
+        for effect in self.input_effects:
+            if effect.cond:
+                manual = True
```
why do you call this 'manual'?

gvanrossum replied:
Because the PUSH() and POP() operations are executed "manually" rather than using TOP() and friends combined with STACK_GROW()/SHRINK(). It's not a great name but I couldn't think of something better that wasn't a whole sentence.

gvanrossum committed:
Fix two bugs in case generator

- UndefinedLocalError when generating metadata for an 'op'
- Accidental newline inserted in test_generator.py

gvanrossum committed:
Merge branch 'fix-op-metadata' into cond-cases

gvanrossum committed:
Blackify

iritkatriel commented on Tools/cases_generator/test_generator.py:
```diff
@@ -476,3 +475,26 @@ def test_register():
+            output = spam(oparg, input);
+            POKE(1, xx);
+            if (oparg & 2) { PUSH(output); }
+            PUSH(zz);
```
Could we stick with PEEK and POKE and here have something like this?

```suggestion
            POKE((oparg & 2) + 2, xx);
            if (oparg & 2) { POKE((oparg & 2) + 1, output); }
            POKE(1, zz);
```

gvanrossum replied:
Interesting. This would favor cases where the conditional is at the bottom of the inputs (e.g. `a if (...), b, c`). But I think it would become quite horrible if you have several conditions, like we have in MAKE_FUNCTION.

iritkatriel replied:
If there are several conditions then the concatenation of the conditions (and the unconditional 1's) gives you the index. 
The generator will be simpler, the generated code will have these expressions for indices, but I think that's probably easier to follow than a mixture of pushes and pokes.

gvanrossum replied:
I'll play around with that.

gvanrossum replied:
Have a look. The generator sure is a lot simpler, I think I can live with the more complicated generated code. What made this solution so nice is that we already had the symbolic size -- the condition just gets added into that. Thanks for the nudge!

gvanrossum committed:
Merge remote-tracking branch 'origin/main' into cond-cases

gvanrossum committed:
Use Irit's idea

gvanrossum commented on Tools/cases_generator/generate_cases.py:
```diff
@@ -59,7 +61,10 @@ def effect_size(effect: StackEffect) -> tuple[int, str]:
     At most one of these will be non-zero / non-empty.
     """
     if effect.size:
+        assert not effect.cond, "Manual effects should be conditional"
```
This message is incorrect.

gvanrossum commented on Tools/cases_generator/generate_cases.py:
```diff
@@ -158,11 +168,21 @@ def declare(self, dst: StackEffect, src: StackEffect | None):
 
     def assign(self, dst: StackEffect, src: StackEffect):
+        if "boerenkool" in dst.name or "boerenkool" in src.name:
+            breakpoint()
```
Debug output accidentally left in.

gvanrossum commented on Tools/cases_generator/generate_cases.py:
```diff
@@ -158,11 +168,21 @@ def declare(self, dst: StackEffect, src: StackEffect | None):
+        #     if src.cond:
+        #         self.emit(f"if ({src.cond}) {{ PUSH({cast}{src.name}); }}")
+        #     else:
+        #         self.emit(f"PUSH({cast}{src.name});")
```
Delete commented-out block.

gvanrossum commented on Tools/cases_generator/generate_cases.py:
```diff
@@ -320,6 +354,7 @@ def write(self, out: Formatter) -> None:
         else:
             # Write output register assignments
             for oeffect, reg in zip(self.output_effects, self.output_registers):
+                src = StackEffect(reg, "")
```
Don't need this.

gvanrossum reviewed (commented):
Notes to self.

gvanrossum committed:
Cleanup

gvanrossum committed:
Change order in LOAD_ATTR output effect

gvanrossum commented on Python/generated_cases.c.h:
```diff
@@ -1788,20 +1782,22 @@
-            JUMPBY(INLINE_CACHE_ENTRIES_LOAD_ATTR);
+            STACK_GROW(((oparg & 1) != 0));
+            POKE(1, res);
+            if (oparg & 1) { POKE(1 + ((oparg & 1) != 0), res2); }
```
It's unfortunate that all new versions test for `oparg & 1` several times, whereas the original version tested that condition exactly once.

iritkatriel replied:
I assumed the compiler would do this, but we could assign ((oparg& i) != 0) to something. I think it’s only used as a condition in one place, in other places it’s just int arithmetic.

gvanrossum committed:
Fix type annotations to satisfy pyright (and mypy mostly)

gvanrossum commented on Python/bytecodes.c:
```diff
@@ -51,7 +51,7 @@
 // Dummy variables for stack effects.
 static PyObject *value, *value1, *value2, *left, *right, *res, *sum, *prod, *sub;
-static PyObject *container, *start, *stop, *v, *lhs, *rhs;
+static PyObject *container, *start, *stop, *v, *lhs, *rhs, *res2;
```
We will also need `static int res2_exists`, and a bunch of other things, before I land this.

gvanrossum committed:
Solve uninitalized variable warning using = NULL

gvanrossum committed:
Use 'X ? 1 : 0' instead of 'X != 0'

This hopefully avoids a silly compiler warning.
